docs/Binding-and-project-context-evaluation.md

```diff
@@ -0,0 +1,73 @@
+The feature is available for both `dotnet new` and Visual Studio.
+
+## `bind` symbols
+
+The symbol binds value from external sources. 
+By default, the following sources are available:
+- host parameters - parameters defined at certain host. For .NET SDK the following parameters are defined: `HostIdentifier: dotnetcli`, `GlobalJsonExists: true/false`.  Binding syntax is `host:<param name>`, example: `host:HostIdentifier`. 
+- environment variables - allows to bind environment allowed. Binding syntax is `env:<environment variable name>`, example: `env:MYENVVAR`. 
+
+It is also possible to bind the parameter without the prefix as the fallback behavior: `HostIdentifier`, `MYENVVAR`.
+
+|Name|Description|
+|---|---|
+|`type`|`bind`|
+|`binding`| Mandatory. The name of the source and parameter in the source to take the value from. The syntax follows: `<source prefix>:<parameter name>`.|
+|`replaces`|The text to be replaced by the symbol value in the template files content.|
+|`fileRename`|The portion of template filenames to be replaced by the symbol value.| 	 
+|`defaultValue`|The value assigned to the symbol if no value was provided from external source(s).|
+
+ 
+##### Example  
+
+```json
+"symbols": {
+   "HostIdentifier": {
+      "type": "bind",
+      "binding": "host:HostIdentifier"
+    }
+}
+```  
+
+### Binding to MSBuild properties
+
+It is possible to bind symbols to MSBuild properties of the current project. The prefix to be used: `msbuild` and it is mandatory.
+Commonly used with item templates to get the information about the project it is added to.
+
+`dotnet new` attempts to find the closest project file using following rules:
+- the project in current directory or `--output` directory (matching `*.*proj` extension)
+- if not found, the parent of above and so on
+In case of ambiguity, the path to the project can be specified using `--project` option.
+
+Once project is located, its MSBuild properties are evaluated. The project should be restored, otherwise evaluation fails.
+Only .NET [SDK-style projects](https://docs.microsoft.com/en-us/dotnet/core/project-sdk/overview) are supported.
+
+It is recommended to configure `defaultValue` for `bind` symbol that will be used in case evaluation fails. 
+If applicable, it is also recommended to use [project capability constraint](https://github.com/dotnet/templating/wiki/Constraints#Project-capabilities) to define the projects that the template can be added to.
+
+Example - binds `DefaultNamespace` symbol to `RootNamespace` of the project:
+```json
+  "symbols": {
+    "DefaultNamespace": {
+      "type": "bind",
+      "binding": "msbuild:RootNamespace",
+      "replaces": "%NAMESPACE%",
+      "defaultValue": "TestNamespace"
+    }
+  },
+  "constraints": {
+    "csharp-only": {
+      "type": "project-capability",
+      "args": "CSharp + TestContainer" // only allowed in C# test project
+    }
+  }
+```
+
+
+## Visual Studio specifics
+TBD
```

Comment:
@phenning @sayedihashimi could you please help to fill this section?
iirc, Visual Studio has another source for bind symbols which would be good to include

Reply:
We have the same msbuild source as well, which performs the same as the cli source

There is also "context" source:
"context:cratesolutiondirectory", - indicates if a solution directory is to be created as a result of project creation (Place solution and project in same directory is UNCHECKED in NPD)

"context:isexclusive" - is the template instantiation is a result of a new project being created (true) vs adding to an existing solution (false)

"context:solutionname" - the name of the solution, which may be different from the project name

Reply:
We will also provide a bind to "namespace" via either:

"type": "bind"
"binding": "namespace"

or Via the  overriding of TryGetHostParamDefault from our DefaultTemplateEngineHost subclass.
This is also how we provide the HostIdentifier, by returning base.TryGetParamDefault when paramName != "namespace".